### ISSUE TITLE
Project Level 3 (Stuck on notion of time)

### ISSUE THREAD

**Issue by Alexius0123** [nus-cs2030/2122-s2#351]
Hello! So far I have been progressing with the project and I have come to a sort of dead-end when it comes to time. So far, the issue that I face is that the notion of time is not being tracked properly within the project.

Refering to the screenshot, at 1.500, the list is not properly updated to show that the servers are available for waiting. This might be due to the logic of my Simulate3 class, which first creates events for the list of customer arrival times, then loops up to 4 times to execute the events for the arrival times before moving on to the next one. 

My question is: Is there a problem with my simulate logic or my event logic? As well as does anyone have a suggestion on how to track the notion of time?

![image](https://user-images.githubusercontent.com/76044965/163677955-8ba0d866-12bf-4ec8-970f-22d40fc82782.png)

**Comment by shotnothing** [comment]
A bit sparse on details, how are you tracking who is waiting on each server?

One way to track the notion of time is let the event loop handle all the timing. The server has a queue of customers (with no explicit reference to time except customers are sorted by arrival times). All you need from the server is

- can server serve?
- can server wait?
- queue up
- leave queue

and a few more in the later levels, but by and large the events are enough to handle all the timing i.e. serve() adds a done() event registered at time+1 units later.

This is only one of the ways to do it but I found it fairly elegant and robust.

**Comment by Alexius0123** [comment]
> A bit sparse on details, how are you tracking who is waiting on each server?
> 
> One way to track the notion of time is let the event loop handle all the timing. The server has a queue of customers (with no explicit reference to time except customers are sorted by arrival times). All you need from the server is
> 
> * can server serve?
> * can server wait?
> * queue up
> * leave queue
> 
> and a few more in the later levels, but by and large the events are enough to handle all the timing i.e. serve() adds a done() event registered at time+1 units later.
> 
> This is only one of the ways to do it but I found it fairly elegant and robust.

Thank you! Does that mean my Server class should make use of a PQ to track customers assigned to them? i.e. 1 server class tracking a maxiumum of 2 customers at once.

**Comment by nyiwei** [comment]
Hello!!  @shotnothing I was able to add a done event registered at time +1 units later however I am a little stuck on how to track the done time to determine the service time for the customer waiting.
Thank you!!

**Comment by nyiwei** [comment]
> > A bit sparse on details, how are you tracking who is waiting on each server?
> > One way to track the notion of time is let the event loop handle all the timing. The server has a queue of customers (with no explicit reference to time except customers are sorted by arrival times). All you need from the server is
> > 
> > * can server serve?
> > * can server wait?
> > * queue up
> > * leave queue
> > 
> > and a few more in the later levels, but by and large the events are enough to handle all the timing i.e. serve() adds a done() event registered at time+1 units later.
> > This is only one of the ways to do it but I found it fairly elegant and robust.
> 
> Thank you! Does that mean my Server class should make use of a PQ to track customers assigned to them? i.e. 1 server class tracking a maxiumum of 2 customers at once.

For me I used a ImList to track the Customers!

**Comment by shotnothing** [comment]
@Alexius0123 yup that's what I did, but PQ or ImList works fine. The canServe and canWait methods basically operate on comparing to the length/size of the PQ/ImList and help gatekeep the flow of customers into the server.

**Comment by shotnothing** [comment]
hi! So the idea is that the `done` event calls the `serve` event for the next customer.

![image](https://user-images.githubusercontent.com/16388604/163688148-63d16a5d-8b93-4f50-8b9a-1d8417c06e2d.png)

**Comment by joannelxl** [comment]
> hi! So the idea is that the `done` event calls the `serve` event for the next customer.
> 
> ![image](https://user-images.githubusercontent.com/16388604/163688148-63d16a5d-8b93-4f50-8b9a-1d8417c06e2d.png)

Thanks @shotnothing! I tried this method, but it is giving me cyclic dependency issues. I'm guessing its because of  the order: 
 `Serve -> `Done' -> `Serve' ?

Do you know how i can solve this issue? Thanks !

**Comment by wannyu** [comment]
Hi! I think you can try calling for the Serve event in Simulate3's run() instead. I made it such that the server keeps track of the waiting customer, so when its done serving the first customer, it creates a new Serve for the second customer.

**Comment by shotnothing** [comment]
> Hi! I think you can try calling for the Serve event in Simulate3's run() instead. I made it such that the server keeps track of the waiting customer, so when its done serving the first customer, it creates a new Serve for the second customer.

yup! that's what I did

**Comment by shotnothing** [comment]
> Hi! I think you can try calling for the Serve event in Simulate3's run() instead. I made it such that the server keeps track of the waiting customer, so when its done serving the first customer, it creates a new Serve for the second customer.

Yup! That's what I did too

**Comment by tristky** [comment]
Hi, I am facing the same issue and after reading the visualisation, i understand much better, thank you so much

**Comment by fionaxjy** [comment]
Hi, I faced the same issue, it was because I did not empty the wait list after moving the customer to the serve event!
